src/server/pps/server/s3g_sidecar.go

```diff
@@ -0,0 +1,359 @@
+		)
+	}()
+	masterLock := dlock.NewDLock(s.apiServer.env.GetEtcdClient(),
+		path.Join(s.apiServer.etcdPrefix,
+			s3gSidecarLockPath,
+			s.pipelineInfo.Pipeline.Name,
+			s.pipelineInfo.Salt))
+	ctx, err := masterLock.Lock(s.pachClient.Ctx())
+	if err != nil {
+		return err
+	}
+	// pachClient := s.pachClient.WithCtx(ctx)
+	defer masterLock.Unlock(ctx)
+	// Watch for new jobs & create kubernetes service for each new job
+	s.HandleJobs(JobHandler{
+		Begin: func(ctx context.Context, jobInfo *pps.JobInfo) error {
+			// Create kubernetes service for the current job ('jobInfo')
+			rcName := ppsutil.PipelineRcName(jobInfo.Pipeline.Name, jobInfo.PipelineVersion)
+			labels := map[string]string{
+				"app":       rcName,
+				"suite":     "pachyderm",
+				"component": "worker",
+			}
+			service := &v1.Service{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "Service",
+					APIVersion: "v1",
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   rcName,
+					Labels: labels,
+				},
+				Spec: v1.ServiceSpec{
+					Selector: labels,
+					Ports: []v1.ServicePort{
+						{
+							Port: int32(s.apiServer.env.S3GatewayPort),
+							Name: "s3-gateway-port",
+						},
+					},
+				},
+			}
+
+			if err := backoff.RetryNotify(func() error {
+				_, err := s.apiServer.env.GetKubeClient().CoreV1().Services(s.apiServer.namespace).Create(service)
+				return err
+			}, backoff.NewExponentialBackOff(), func(err error, d time.Duration) error {
+				if strings.Contains(err.Error(), "already exists") {
+					return err // success
+				}
+				logrus.Errorf("error creating kubernetes service for s3 gateway sidecar: %v; retrying in %v", err, d)
+				return nil
+			}); err != nil && !strings.Contains(err.Error(), "already exists") {
+				return err
+			}
+			return nil
+		},
+
+		End: func(jobID string) error {
+			if !ppsutil.ContainsS3Inputs(s.pipelineInfo.Input) && !s.pipelineInfo.S3Out {
+				return nil // don't need to delete kubernetes service
+			}
+			rcName := ppsutil.PipelineRcName(s.pipelineInfo.Pipeline.Name, s.pipelineInfo.Version)
+
+			if err := backoff.RetryNotify(func() error {
+				return s.apiServer.env.GetKubeClient().CoreV1().Services(s.apiServer.namespace).Delete(rcName,
+					&metav1.DeleteOptions{OrphanDependents: new(bool) /* false */})
+			}, backoff.NewExponentialBackOff(), func(err error, d time.Duration) error {
+				if strings.Contains(err.Error(), "not found") {
+					return err // success
+				}
+				logrus.Errorf("error deleting kubernetes service for s3 gateway sidecar: %v; retrying in %v", err, d)
+				return nil
+			}); err != nil && !strings.Contains(err.Error(), "not found") {
+				return err
+			}
+			return nil
+		},
+	})
+	return nil
+}
+
+type JobHandler struct {
+	// Begin runs when a job is created
+	Begin func(ctx context.Context, jobInfo *pps.JobInfo) error
+
+	// End runs when a job ends
+	End func(jobID string) error
+}
```

Comment:
It seems to me like it'd be more idiomatic to use an interface here

Reply:
Found it! Sorry, I think this comment was under at least two "show hidden" buttons. I did this refactoring, but it got a little bit strange because there are some methods of `jobHandler` that are meant to be different across the two implementations (`OnCreate` and `OnTerminate`) and some that aren't (`start`, `processJobEvent`, etc), and both access many of the same variables.

I just finished this refactoring and will push it so you can see what it looks like. If you think it's better we can keep it, but if not it's easy to drop. Just let me know.